octopus_sensing/tests/test_camera_streaming.py

```diff
@@ -0,0 +1,134 @@
+
+    '''
+    Get the information of your computer using:
+    >>> import cv2
+    >>> cap=cv2.VideoCapture(0)
+    >>> print(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+    >>> print(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+    '''
+    def CAP_PROP_FRAME_WIDTH(self):
+        return (float(640.0))
+    
+    def CAP_PROP_FRAME_HEIGHT(self):
+        return (float(480.0))
+    
+    def imwrite(self, filename, img):
+        a = time.time()
+        # self.output_path/self.name-experiment_id-str(message.stimulus_id).zfill(2)/str(a).jpg
+        open(filename[:-4] + "/" + str(a) + ".jpg", 'w+').write(img)
+
+class MockedVideoCaptureModule:
+
+    def isOpened(self):
+        return True
+
+    def set(self):
+        pass
+
+    def read(self):
+        # bool, 3d matrix
+        return True, [[[0,10,10],[1,11,11]], [[0,10,10],[1,11,11]]]
+    
+    def release(self):
+        pass
+
+    def get(self):
+        return (float(640.0))
+
+
+class MockedVideoWriterModule:
+
+    def __init__(self, file_name):
+        self.file_name = file_name
+
+    def write(self, frame):
+        a = time.time()
+        print('self.file_name:',self.file_name) # used for bug checking
+        open(self.file_name, 'w+').write(frame)
+
+    def release(self):
+        pass
+
+@pytest.fixture(scope="module")
+def mocked():
+    original_bases = camera_streaming.CameraStreaming.__bases__[0].__bases__
+    camera_streaming.CameraStreaming.__bases__[0].__bases__ = (threading.Thread,)
+
+    original_cv2 = camera_streaming.cv2
+    camera_streaming.cv2 = MockedCv2Module()
+
+    yield None
+
+    # Replacing back the original things
+    camera_streaming.CameraStreaming.__bases__[0].__bases__ = original_bases
+    camera_streaming.cv2 = original_cv2
+
+
+def test_happy_path(mocked):
+
+    output_dir = tempfile.mkdtemp(prefix="octopus-sensing-test")
+    device_name = 'test-video-device'
+    experiment_id = 'test-exp-1'
+    stimuli_id = 'sti-1'
+
+    device = camera_streaming.CameraStreaming(
+        camera_no=0, output_path=output_dir, name=device_name)
+
+    msg_queue = queue.Queue()
+    device.set_queue(msg_queue)
+    device.start()
+
+    time.sleep(0.2)
+
+    msg_queue.put(start_message(experiment_id, stimuli_id))
+
+    time.sleep(1)
+
+    msg_queue.put(stop_message(experiment_id, stimuli_id))
+
+    time.sleep(0.2)
+
+    # It should save the file after receiving a STOP.
+    device_output = os.path.join(output_dir, device_name)
+    assert os.path.exists(device_output)
+
+    recorded_file = os.path.join(device_output, '{}-{}-{}.avi'.format(
+        device_name, experiment_id, stimuli_id))
+
+    
+    ''' 
+    for bug testing only
+    list_dir = os.listdir(device_output)
+    print ('output_dir:',output_dir)
+    print('device_name:',device_name)
+    print ('device_output:', device_output)
+    print("list_dir:",list_dir)
+    for temp in list_dir:
+        print("temp:",temp)
+    '''
+
+    assert os.path.exists(recorded_file)
```

Comment:
While I was running the test, I thought, we should have an assertion here for the file size too. To ensure that the device actually wrote something to the file:

```
assert os.path.getsize(recorded_file) > 102400
```

Reply:
Thanks for these helpful suggestions! I've fixed those bugs and pushed the latest file here. Glad if you could check the latest file and see if there is anything wrong!

I make a small change in checking the file size. Originally the fake signal is a 2*2 matrix, and when I ran the code, the error appears:
`AssertionError: assert 38864 > 102400`
When I change the fake signal into a 3*3 matrix:
`return True, [[[0,10,10],[1,11,11],[1,11,11]], [[0,10,10],[1,11,11],[1,11,11]],[[0,10,10],[1,11,11],[1,11,11]]]`
the error disappears and it works fun. So I'm thinking maybe the previous error is because 102400 is too big. So I keep the former matrix and change 102400 into 25600. I'm not sure whether it's right to do so.

Reply:
On my machine, it was producing a larger file! I guess it's a matter of how fast or slow the test in running. Any number is okay. All is good with me.

Thanks for all the hard work! I'll merge your changes now.

Reply:
All tests passed on our CI too :tada: 
https://github.com/octopus-sensing/octopus-sensing/actions/runs/1895863905

Reply:
Thanks so much for your help! I'm very glad about it😄